### PR TITLE
[Do Not Merge] Obscuring Reagents

### PR DESCRIPTION
**Pull request by Deantwo**
I was reading around some of the older issues and found this: https://github.com/tgstation/-tg-station/issues/5547#issuecomment-61203224. That along with having a game where I could easily tell from afar that a traitor chemist was making _bad_ stuff made me want to try this.
I though up this little idea about making you only able to see the individual chemicals while wearing either **Science Goggles** or a **Health Scanner HUD**. Because making it so that a Chemist has to use his PDA to scan all the beakers he works with would be a total nightmare.

**Changes:**
- If you are not wearing either Science Goggles or a Health Scanner HUD, you can no longer see what types of reagents are in a beaker/bottle/etc. This only works up close though.
- Added a new variable to glasses with a hud.dm define flag, so it is easy to allow other types of glasses to use this feature.
- You can no longer see how much is in a non-transperent container from afar. (Buckets, etc.)
- Added a new define flag, TRANSPARENT, to all items so it is easy to allow OPENCONTAINER items to disable this feature.

**Picture examples:** (a little outdated, but the idea still stand)
reagentscanner without glasses reagentscanner with glasses

reagentscanner bucket

**Needed feedback:**
- I don't have a that good knowledge of what else actually could use this or what I may have forgotten.
- Should **Science Goggles** be renamed with this?
- ~~Should it be described as "liquid" instead of "chemicals"?~~
- ~~Is the "You cannot see what it contains from here" message too long?~~
- Should other types of items be able to do this then just glasses? Maybe helmets? Gloves?
- Further ideas

**Things to do:**
- [ ] Handheld Reagent Scanner (report all reagents in target, highlight bad reagents? highlight if over overdose threshold? upgrade PDA reagent scanner to the same?)
- [ ] Make beakers be the only reagent container that give exact unit amounts
- [ ] Ask community more about what they want/think (make a feedback forum thread)

### CODE REVIEW

**Comment by Jordie0608** [comment]
Cool idea and finally a use for science goggles. :+1: 

Does it affect the bartender's drinks?

**Comment by Deantwo** [comment]
> Does it affect the bartender's drinks?

As far as I know you can't see what is in the bartender's drinks already. And I didn't change any of the default behavior other then making it say ```"[reagents.total_volume] units of chemical"``` instead of listing all the reagents.

**Comment by Deantwo** [comment]
I so wanna fill the janitor's bucket and/or janitorial cart with welding fuel.

**Comment by Jordie0608** [comment]
>That's The Manly Dorf.
>A manly concotion made from Ale and Beer. Intended for true men only.
>It contains:
>36 units of The Manly Dorf
>8 units of Beer
>1 units of Ale
>It is a tiny item.

**Comment by Deantwo** [comment]
Is that with or without my changes?

**Comment by Jordie0608** [comment]
Without

**Comment by Deantwo** [comment]
Then with my change it will say:

>That's The Manly Dorf.
A manly concotion made from Ale and Beer. Intended for true men only.
It contains:
55 units of  chemical
It is a tiny item.

**Comment by Deantwo** [comment]
<Deantwo> would like it to say whether the chemicals/liquids in a bucket are clear or colored though... not sure how
<Jordie> could be snowflakey
<Jordie> or tedious
<Deantwo> i mean... you can see the diffence between welding fuel and water... yes?
<Deantwo> yeah... -_-
<Deantwo> maybe adding a var to all chems... with a kinda of "look" description
<Jordie> so you'd have "40 units of a oily chemical"?
<Deantwo> but then would have to make a piority var too so some descriptions can override others
<Deantwo> yeah
<Jordie> or just go with whatever has the most
<Deantwo> yeah...
<Jordie> entirely possible, but again: tedious

**Comment by optimumtact** [comment]
:+1:

File: code/game/atoms.dm [hunk 1]
```diff
@@ -246,12 +249,19 @@ its easier to just keep the beam vertical.
...
+				if(G && (G.scan_vision & SCAN_CHEMICAL) && get_dist(user, src) <= 1)
+					for(var/datum/reagent/R in reagents.reagent_list)
+						user << "[R.volume] units of [R.name]"
+				else
+					user << "[reagents.total_volume] units of chemical" // Water is a chemical, right?
+			else
+				user << "Nothing."
 		else
-			user << "Nothing."
+			user << "You cannot see what it contain from here."
```
Comment (RemieRichards) [inline code/game/atoms.dm]: contains.

Comment (Deantwo) [inline code/game/atoms.dm]: Thank. Trying to think of another way of wording it tough.

Comment (Jordie0608) [inline code/game/atoms.dm]: "You can't make out the contents"?

**Comment by Lumipharon** [comment]
Yes please, merge post haste.
Not being able to spike people's drinks after someone made the contents of glasses visible was pretty gay.

I would suggest making it ONLY science goggles though, since
A: if both work, then science goggles are just shittier medhuds
B: medhuds already have their own purpose/functions.

**Comment by Deantwo** [comment]
Yeah, I mostly added it to Health Scanner HUD as well because medical doctors will have no way to scan the content without it.

My second idea would be to merge the ChemWhiz and Med-U cartridges. But Med-U already have a scanner feature, so changing between the two scanners would be really annoying.

So do we wanna totally blind Medical Doctors too?

**Comment by Lumipharon** [comment]
Put a couple of analysers or some shit in med storage, so they can scan shit when needed.

**Comment by Deantwo** [comment]
> Put a couple of analysers or some shit in med storage, so they can scan shit when needed.

Can analyzers scan reagents?

**Comment by Lumipharon** [comment]
I  can't remember. Maybe not. Otherwise I think the Mass-Spectrometer (currently a low level R&D item) can.
And if neither can, then well fuck, create something new/add that functionality.

**Comment by AlexCrimson** [comment]
Could you not add constructable Science Googles/Night Vision Science Goggles to the Protolathe? Could be a research thing. We have all the other HUDs in there.

+1 for addressing this issue. Now we will be able to spike drinks again!

Only certain PDA cartridges can scan reagents in a container. The Spectrometer scans blood that you inject into it. No idea about the Health Analyzer, i assume it doesnt scan reagents.

**Comment by Deantwo** [comment]
Yeah only the ChemWhiz and Med-U Deluxe cartridges can scan reagents.
So maybe a handheld device would should be added too.

**Comment by Dorsisdwarf** [comment]
1) Just like goon! Yay! Everyone needs more goon in their station! 

2) Adds a handheld device which removes the uniquity of the secondary function from the forensic scanner, the only thing making the Detective a relevant job as opposed to Sec Officer 2: The Private Eye

**Comment by Kromgar** [comment]
Detectives scanned for chemicals in glasses all the time Doris?

**Comment by Deantwo** [comment]
Wait Goon has this already? And here I thought I had an awesome idea that no one had thought about.

Anyway, what are you talking about with the detective? He can scan for reagents? Didn't even know that.
I haven't actually started on the handheld device yet, so if you have a better idea then do share.

**Comment by Profakos** [comment]
We had this on Sayustation for a few years, including science and medhud's ability to detect reagents, at first I thought it would be annoying, but it worked out.

**Comment by Cheridan** [comment]
>So far at least two suggest it changed to "liquid" instead of "chemicals".

Not all reagents are liquids though. Some are gases or solid powders. Otherwise it's a good premise.

**Comment by Deantwo** [comment]
>> So far at least two suggest it changed to "liquid" instead of "chemicals".

> Not all reagents are liquids though. Some are gases or solid powders. Otherwise it's a good premise.

Like I was [talking with Jordie about](https://github.com/tgstation/-tg-station/pull/7402#issuecomment-71807372). Best way would be to add a var to each chem that can be used to describe what it looks like and then use the reagent that is dominant in the beaker.
Thoughts?

**Comment by Incoming5643** [comment]
my thought is to define various color descriptor by ranges on RGB then have that appear in the description. Then you could add modifiers for both the opacity and phase state of the chemicals. This is a lot of work though so if you want to go with something simpler I might just code this mysellf at some point.

**Comment by AlexCrimson** [comment]
Is it really that important? Chemical, Liquid, Reagent, Substance, Compound. Do the little details really matter enough to do all that annoying code?

**Comment by Deantwo** [comment]
Yeah was thinking about doing the RGB code too. Just not sure how I would do it. This would at least be useful for beaker that don't have visible content.
Kinda like the [little colored thing that is on the wiki right now](https://tgstation13.org/wiki/Guide_to_chemistry#Medicines).

> Is it really that important? Chemical, Liquid, Reagent, Substance, Compound. Do the little details really matter enough to do all that annoying code?

It may not need to be done right away. But at some point.

File: code/game/atoms.dm [hunk 2]
```diff
@@ -246,12 +249,19 @@ its easier to just keep the beam vertical.
...
-				user << "[R.volume] units of [R.name]"
+		if(get_dist(user, src) <= 1 || is_transparent_container())
+			user << "It contains:"
+			if(reagents.reagent_list.len)
+				var/obj/item/clothing/glasses/G = user.get_item_by_slot(slot_glasses)
+				if(G && (G.scan_vision & SCAN_CHEMICAL) && get_dist(user, src) <= 1)
+					for(var/datum/reagent/R in reagents.reagent_list)
+						user << "[R.volume] units of [R.name]"
+				else
+					user << "[reagents.total_volume] units of chemical" // Water is a chemical, right?
```
Comment (Deantwo) [inline code/game/atoms.dm]: Gonna post this here.
So far the idea have been to change it to "liquid".
Another idea could be "reagents".
Sadly finding one word that good for both chemistry and drinks.

File: code/game/atoms.dm (hunk 1 above)
Comment (Deantwo) [inline code/game/atoms.dm]: Could just go with the same as paper.
"<span class='notice'>It is too far away.</span>"

File: code/game/atoms.dm (hunk 2 above)
Comment (Miauw62) [inline code/game/atoms.dm]: Can't you just do "It contains [amount] units of something, but you can't quite make out what"

Comment (Deantwo) [inline code/game/atoms.dm]: Oh that is a good idea!
I feel stupid for not thinking of that.

**Comment by KazeEspada** [comment]
Reagents have a variable known as Reagent State which states whether they are Solids, Liquids, or even Gases. Its a actually a number defined at the top of Chemistry-Reagents.dm. You could use that to show whether the beaker is filled with a liquid, solid, or gas.

**Comment by Deantwo** [comment]
Ah cool.
I'll look at it tomorrow.

**Comment by Fox-McCloud** [comment]
Goon did just this; they make "spectroscopic" goggles; if you're wearing them, you can tell exactly what's in a beaker, no matter the distance.

If you're not wearing them, it'll just say what color the chemical is "it's a reddish-brown chemical"

This applies to transparent or non-transparent containers though, which I kinda agree with. Yes, I realize it's "realism", but this is just going to encourage every chemist and their brother to ask for buckets and other containers all the time in an effort to not get caught (then get metagamed every time they ask for one too).

**Comment by PKPenguin321** [comment]
My only request is the bartender gets some science goggles so I can still measure my drinks out without having to guess every time I'm interrupted when I'm mixing something. That, or make it so anyone can see what reagents are in the shaker without needing goggles.

**Comment by Iamgoofball** [comment]
imo give the bartender a custom sprite subtype of the science goggles so he doesn't look silly with the goggles

**Comment by Antimattercarp** [comment]
Green teashades for the bartender?

**Comment by An0n3** [comment]
Bartender should have an item called "beer goggles" so he can keep an eye on his drink mixes (for the terrible bartenders who can't just remember).  If these could also somehow measure the BAC of people he looks at that would be hilarious.  YOU'VE BEEN CUT OFF

**Comment by paprka** [comment]
>BEER GOGGLES

PLEASE DO THIS

BEST SUGGESTION

**Comment by PKPenguin321** [comment]
redirect beer goggles discussion to here please since I feel bad about sparking off-topic comments in Deantwo's PR: https://tgstation13.org/phpBB/viewtopic.php?f=9&t=2763

**Comment by Tenebrosity** [comment]
Great PR

An0n3's suggestion is pretty good too

**Comment by Deantwo** [comment]
> Prevent bottles from being refilled due to the narrowness of their mouths and the lack of funnels on the station (Bartender currently starts with 25+ free 100u beakers)

Currently making a stop to this in another planned PR. Hopefully see that done tomorrow.

As for this PR, I am a little unsure what people would want done.
Do we need a real forum discussion thread about this? With polls?
If someone wants to make the handheld reagent scanner then go right ahead.

**Comment by Lumipharon** [comment]
Forum poll/thread nver hurts, but as for not being able to refill bottles - 
Have you guys never done this in real life? Sure, I wouldn't want to try it with polyacid, but it's not hard to put shit in a bottle without spilling, especially if you are doing it from a beaker, which have narrow lips that make it easier.

How would you even 'exploit' filling old vodka bottles and shit anyway? You can't use them in nades (afaik), can't splash them on people, you can only force feed or headsmash.

**Comment by Tenebrosity** [comment]
I suggested it because I thought it would be consistent with the idea behind this pull. But if it's going to be a different pull, we should talk about it there.

**Comment by Iamgoofball** [comment]
yeah save the beer bottle changes for a diff. pr

**Comment by Deantwo** [comment]
I'd say keep this PR here so we have a goal with it.
But a number of things need done before this is a good idea.

Still, feel free to create the required devices in other PRs.
For example the handheld reagent scanner needs to be a thing.

**Comment by Jordie0608** [comment]
Changelog.

**Comment by SuperSayu** [comment]
I did almost exactly this for Sayustation, the most notable differences being that when only a few reagents were there you could identify them (meaning it became unreadable when mixed) and that beaker contents reporting was split between solid, liquid, and gas (you could identify one gas, but a mixture of three liquids was confusing).

Players *hated* it and it was dialed back but never eliminated.

My favorite consequence of the system was allowing mystery chembottles.  I made a whole box full of em that you could order in cargo.

I also gave completely screwed up results if you were drugged while looking at beakers, which is fun (this beaker is full of 110 units of spiders!).

**Comment by Cheridan** [comment]
@Deantwo Is progress still being made on this?

**Comment by Deantwo** [comment]
I haven't really had the time no.
Mostly stuck on trying to figure out what people actually want.

People didn't seem to like the idea I made for the bucket at least.

**Comment by Razharas** [comment]
Last activity on second of march, i guess this is dead
If its not @Deantwo please comment on the progress

**Comment by Deantwo** [comment]
Yeah, gotten rather side tracked.
Never figured out exactly what people wanted.

If someone wanna take it over they can, or if people can decide.

Most of all it is waiting on https://github.com/tgstation/-tg-station/pull/7657, which needs the most work done.

**Pull request closed without merging**